designs/full-cluster-lifecycle-api.md

```diff
@@ -0,0 +1,224 @@
+
+After a management cluster is created, user can define a workload cluster spec with the option to specify a custom namespace to deploy the EKS-A resources (all resources need to be in the same namespace), and directly execute `kubectl apply/create -f workload-spec.yaml` in the management cluster or push the workload cluster spec to the Git repo when GitOps is enabled, to create a workload cluster. Once the cluster is in ready state, user is able to fetch the workload cluster `kubeconfig` from the management cluster through `kubectl`, or by running a CLI command.
+
+User can then manage a workload cluster lifecycle in the native Kubernetes way by editing its resources through `kubectl edit/patch/delete` command, or through Git when the GitOps is enabled on management cluster.
+
+## Security
+
+There are three main changes in terms of security:
+
+1. The EKS-A controller will have full control over CAPI objects (which gives full access to the the underlying cluster infrastructure)
+2. The EKS-A controller will have ClusterAdmin access to all workload clusters. From a role access perspective the current controller already has access to the secrets containing the kubeconfigs for the workload clusters. But there is no code to retrieve and use them.
+3. Any user with the ability to create EKS-A (in any namespace) objects will be able to create clusters
+
+There is very little we can do about the first one since it’s required by the very own nature of this design. We should keep making sure we give the users the tools and information to secure the controller.
+
+The second one could palliated by using a different user than only has the permissions that the controller needs (to mention some: install the CNI, update vSphere controllers and DS, update CoreDNS permissions...). However, this user will need be created at some point using the admin credentials, since those are the only ones provided by CAPI. So in fact, this will just pushing the risk somewhere else.  An alternative could be to add a new feature to CAPI that allows us to specify optional users/roles to be created automatically.
+
+Both the controller and the vSphere credentials live in the `eksa-system` namespace. RBAC can be used to secure it. We will document that this namespace should only be accessible by `ClusterAdmin` (or some other high level access role).
+
+In order to mitigate the third one we will add the ability to specify one (or a group) of namespaces that the controller will watch. This will reduce the attack vector, allowing cluster administrators to secure those namespaces accordingly. However this feature is not in this design’s scope and will be designed and delivered separately.
+
+It’s worth noting that, one added benefit of this design is that, workload clusters have very little vSphere access. So Ops teams can give `ClusterAdmin` access to the workload clusters to developers without risking exposing vSphere admin credentials.
```

Comment:
👍  but isn't this the case for clusters that are not self managed anyway?

Reply:
You are right. I'll rephrase this.
I think what I missed here (and it's probably the most important thing) is that with the current implementation to create workload clusters (through the CLI) you still need "admin" access to the management cluster (not really admin but basically total access to the `eksa-system` namespace and CAPI resources). And that gives you access to the vsphere credentials and to any other cluster for that matter, including the management cluster.

Reply:
Done